### DISCUSSION TITLE
Adding new device to automatically install latest NodeJS before installing the agent

### DISCUSSION THREAD

**Discussion by chopeta** [SquirrelCorporation/SquirrelServersManager#41] [Ideas]
Would like SSM to automatically install NodeJS (+ any additional requirement) when installing new agents (while adding new devices).

Would that be possible?

I also found this kind of confusing when I started because the documentation mentions this:

![image](https://github.com/SquirrelCorporation/SquirrelServersManager/assets/7503721/b58cb7a2-d4d6-41d2-a05c-3452080c4610)

But I couldn't make a new agent installation to work until I manually installed NodeJS in the device before adding it into SSM UI.

Here's the log with the error when you try to add a new device without NodeJS previously installed:

<code>
Starting...
Using /server/src/ansible/ansible.cfg as config file
PLAY [Install agent on targeted device] ****************************************
TASK [Install agent on targeted device] ****************************************
 ok: [192.168.122.76] => {
  "msg": "Host ID 1880f949-4e20-45a7-bbca-f43d6f5931bc with API URL http://192.168.122.76:8000"
}
TASK [Check if NodeJS does exist] **********************************************
ok: [192.168.122.76] => {"changed": false, "failed_when_result": false, "msg": "non-zero return code", "rc": 1, "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}
TASK [Check if NPM does exist] *************************************************
ok: [192.168.122.76] => {"changed": false, "failed_when_result": false, "msg": "non-zero return code", "rc": 1, "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}
TASK [Install "PM2" node.js package globally.] *********************************
[WARNING]: Platform linux on host 192.168.122.76 is using the discovered Python
interpreter at /usr/bin/python3.11, but future installation of another Python
interpreter could change the meaning of that path. See
https://docs.ansible.com/ansible-
core/2.17/reference_appendices/interpreter_discovery.html for more information.
fatal: [192.168.122.76]: FAILED! => {"ansible_facts": {"discovered_interpreter_python": "/usr/bin/python3.11"}, "changed": false, "msg": "Failed to find required executable \"npm\" in paths: /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"}
PLAY RECAP *********************************************************************
192.168.122.76 : ok=3 changed=0 unreachable=0 failed=1 skipped=0 rescued=0 ignored=0
</code>

This is my first attempt at using Ansible so I am not an expert at all in the subject... but the way I understand, the "_installAgent" playbook currently manage only the " PM2" dependency, so we would have to improve this playback to actually install NodeJS (from the official repo instead of your own linux distribution default packages) and anything else that could be missing.

I don't see this a bug and not sure if you want to add this "feature" so I am reporting it here as an idea.

**Comment by SquirrelDeveloper** [comment]
Yes, your absolutely right, there is also a requirement of the minimum nodejs version that must be installed too to be checked before installing the agent

**Comment by TheOddPirate** [comment]
First time im playing with ansible today so this might be a dumb fix, but after having problems with the first agent setup i edited the _installAgent playbook to this and after that every agent has installed perfectly 

`
- name: Install agent on targeted device
  hosts: all
  become: true
  gather_facts: false
  vars:
    base_path: /opt/squirrelserversmanager
  tasks:
    - name: Install agent on targeted device
      ansible.builtin.debug:
        msg: Host ID {{ _ssm_deviceId }} with API URL {{ _ssm_masterNodeUrl }}



    - name: Check if NodeJS does exist
      ansible.builtin.raw: which node
      check_mode: false
      changed_when: false
      failed_when: which_res_node.rc > 1
      register: which_res_node
    
    - name: Check if NPM does exist
      ansible.builtin.raw: which npm
      check_mode: false
      changed_when: false
      failed_when: which_res_npm.rc > 1
      register: which_res_npm
      
    - name: Install nodejs and npm if not found
      ansible.builtin.package:
        name:
          - npm
          - nodejs
        state: present  
      when: which_res_npm or which_res_node is failed


    - name: Install "PM2" node.js package globally.
      community.general.npm:
        name: pm2
        global: true
        state: latest

    - name: Install PM2 LogRotate
      command: pm2 install pm2-logrotate

    - name: Check out agent
      ansible.builtin.git:
        force: true
        repo: 'https://github.com/SquirrelCorporation/SquirrelServersManager-Agent.git'
        dest: "{{ base_path }}"
      timeout: 600

    - name: Stop PM2 Agent if running
      command: pm2 stop agent
      ignore_errors: yes

    - name: Delete PM2 Agent if present
      command: pm2 delete agent
      ignore_errors: yes

    - name: Clean directory
      command:
        chdir: "{{ base_path }}"
        cmd: rm -f ssm-agent && rm -f agent.blob && rm -rf ./build && rm -f ./hostid.txt

    - name: Write Node Url in .env file
      copy:
        content: "API_URL_MASTER={{ _ssm_masterNodeUrl }}"
        dest: "{{ base_path }}/.env"

    - name: Write HostId in hostid.txt file
      copy:
        content: "{{ _ssm_deviceId }}"
        dest: "{{ base_path }}/hostid.txt"

    - name: NPM install
      command:
        chdir: "{{ base_path }}"
        cmd: npm install

    - name: NPM run build
      command:
        chdir: "{{ base_path }}"
        cmd: npm run build

    - name: Start PM2 Agent
      command:
        chdir: "{{ base_path }}"
        cmd: pm2 start -f "./build/agent.js"

    - name: Install PM2 on startup
      command: pm2 startup

    - name: Save Agent on startup
      command: pm2 save

    - name: Save Agent on startup
      command: pm2 update


`

**Reply by SquirrelDeveloper** [reply]
@n0tOdd It seems you just updated the playbook to install node and npm. The current playbook is built to fail if those packages are not present.
